api4/work_templates.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package api4
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/v6/app/worktemplates"
+	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+)
+
+func (api *API) InitWorkTemplate() {
+	api.BaseRoutes.WorkTemplates.Handle("/categories", api.APISessionRequired(areWorkTemplatesEnabled(getWorkTemplateCategories))).Methods("GET")
+	api.BaseRoutes.WorkTemplates.Handle("/categories/{category}/templates", api.APISessionRequired(areWorkTemplatesEnabled(getWorkTemplates))).Methods("GET")
+	api.BaseRoutes.WorkTemplates.Handle("/execute", api.APIHandler(areWorkTemplatesEnabled(executeWorkTemplate))).Methods("POST")
+}
+
+func areWorkTemplatesEnabled(h handlerFunc) handlerFunc {
+	return func(c *Context, w http.ResponseWriter, r *http.Request) {
```

Comment:
Agree it seems like too much to add a boolean in the context. How about calling within the routes like

```go
func getWorkTemplateCategories(c *Context, w http.ResponseWriter, r *http.Request) {
        enabled := areWorkTemplatesEnabled(c,w,r)
        if !enabled {
            return
        }
	t := c.AppContext.GetT()
```

Reply:
It's doable, I really wanted to avoid this lol.

Reply:
and as a reviewer I'm even more likely to not think about this at all if those lines don't show during my review.

Reply:
I still think we want to use this as a function within the handler rather than within the middleware. I agree that its easier to forget about. I think the concern of breaking monitoring weighs more heavily.

Reply:
ping @agnivade, do we have a recommended way for "custom built" middlwares?

Reply:
The recommended way would be to have a separate function, and just pass Context to it, and let it return an error. At the call site, check if err != nil, and then construct an AppError, set `c.Err = appErr` and return.

The current way of calling `http.NotFound` is actually incorrect. We usually don't call http methods directly, rather set the status code in the appError and let it be handled in the higher layers.

Please follow the way it is done in https://github.com/mattermost/mattermost-server/blob/ed6b9123cdb28b271a830c1bfeda4f75b9b0a2b2/api4/insights.go#L43-L52

Reply:
Sounds good, I'll apply the changes to match this.